codegen/src/test/resources/software/amazon/awssdk/codegen/poet/builder/test-composed-sync-default-client-builder.java

```diff
@@ -0,0 +1,130 @@
+    @Override
+    protected final String serviceEndpointPrefix() {
+        return "json-service-endpoint";
+    }
+
+    @Override
+    protected final String serviceName() {
+        return "Json";
+    }
+
+    @Override
+    protected final SdkClientConfiguration mergeServiceDefaults(SdkClientConfiguration config) {
+        return config.merge(c -> c.option(SdkClientOption.ENDPOINT_PROVIDER, defaultEndpointProvider())
+                                  .option(SdkAdvancedClientOption.SIGNER, defaultSigner())
+                                  .option(SdkClientOption.CRC32_FROM_COMPRESSED_DATA_ENABLED, false)
+                                  .option(SdkClientOption.SERVICE_CONFIGURATION, ServiceConfiguration.builder().build())
+                                  .option(AwsClientOption.TOKEN_PROVIDER, defaultTokenProvider())
+                                  .option(SdkAdvancedClientOption.TOKEN_SIGNER, defaultTokenSigner()));
+    }
+
+    @Override
+    protected final SdkClientConfiguration finalizeServiceConfiguration(SdkClientConfiguration config) {
+        List<ExecutionInterceptor> endpointInterceptors = new ArrayList<>();
+        endpointInterceptors.add(new JsonResolveEndpointInterceptor());
+        endpointInterceptors.add(new JsonEndpointAuthSchemeInterceptor());
+        endpointInterceptors.add(new JsonRequestSetEndpointInterceptor());
+        ClasspathInterceptorChainFactory interceptorFactory = new ClasspathInterceptorChainFactory();
+        List<ExecutionInterceptor> interceptors = interceptorFactory
+            .getInterceptors("software/amazon/awssdk/services/json/execution.interceptors");
+        List<ExecutionInterceptor> additionalInterceptors = new ArrayList<>();
+        interceptors = CollectionUtils.mergeLists(endpointInterceptors, interceptors);
+        interceptors = CollectionUtils.mergeLists(interceptors, additionalInterceptors);
+        interceptors = CollectionUtils.mergeLists(interceptors, config.option(SdkClientOption.EXECUTION_INTERCEPTORS));
+        ServiceConfiguration.Builder serviceConfigBuilder = ((ServiceConfiguration) config
+            .option(SdkClientOption.SERVICE_CONFIGURATION)).toBuilder();
+        serviceConfigBuilder.profileFile(serviceConfigBuilder.profileFileSupplier() != null ? serviceConfigBuilder
+            .profileFileSupplier() : config.option(SdkClientOption.PROFILE_FILE_SUPPLIER));
+        serviceConfigBuilder.profileName(serviceConfigBuilder.profileName() != null ? serviceConfigBuilder.profileName() : config
+            .option(SdkClientOption.PROFILE_NAME));
+        if (serviceConfigBuilder.crossRegionAccessEnabled() != null) {
+            Validate.validState(
+                clientContextParams.get(JsonClientContextParams.CROSS_REGION_ACCESS_ENABLED) == null,
+                "Cross region access enabled has been configured on both ServiceConfiguration and the client level. Please limit this configuration to one location.");
```

Comment:
Is there a reason we support both places? Is it hard to not support it in S3ClientConfiguration?

Reply:
Here's the proofread version of the text:

The reason it is supported in both places is that the existing S3ClientContextParams are available in both locations. To access the values of these context parameters, they are accessed through getters on the S3Configuration.
```java
        if (serviceConfigBuilder.crossRegionAccessEnabled() != null) {
            Validate.validState(Throw Exception in both places);
        } else {
            serviceConfigBuilder.crossRegionAccessEnabled(clientContextParams
                    .get(S3ClientContextParams.CROSS_REGION_ACCESS_ENABLED));
        }
```

What if the user wants to create a Config object solely to set the Client config in that S3 Configuration?"



"Is it difficult to not support it in S3ClientConfiguration?"

"It's not that difficult. We need to pass ContextParams to the decorator along with the service configuration to determine if it's a cross-region request."

"If we do not support this in S3Configuration, we will not be able to group it along with other client settings like useArnRegionEnabled, multiRegionEnabled, etc."

Reply:
Removed we can always add this in future if required